enterprise/wsproxy/keycache.go

```diff
@@ -0,0 +1,252 @@
+	k.keysMu.RUnlock()
+
+	now = k.Clock.Now()
+	if latest.CanSign(now) {
+		return latest, nil
+	}
+
+	_, latest, err := k.fetch(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("fetch: %w", err)
+	}
+
+	return latest, nil
+}
+
+func (k *CryptoKeyCache) Verifying(ctx context.Context, sequence int32) (codersdk.CryptoKey, error) {
+	if k.isClosed() {
+		return codersdk.CryptoKey{}, cryptokeys.ErrClosed
+	}
+
+	k.keysMu.RLock()
+	key, ok := k.keys[sequence]
+	k.keysMu.RUnlock()
+
+	now := k.Clock.Now()
+	if ok {
+		return validKey(key, now)
+	}
+
+	k.fetchLock.Lock()
+	defer k.fetchLock.Unlock()
+
+	if k.isClosed() {
+		return codersdk.CryptoKey{}, cryptokeys.ErrClosed
+	}
+
+	k.keysMu.RLock()
+	key, ok = k.keys[sequence]
+	k.keysMu.RUnlock()
+	if ok {
+		return validKey(key, now)
+	}
+
+	keys, _, err := k.fetch(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("fetch: %w", err)
+	}
+
+	key, ok = keys[sequence]
+	if !ok {
+		return codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	return validKey(key, now)
+}
+
+// refresh fetches the keys from the control plane and updates the cache.
+func (k *CryptoKeyCache) refresh() {
+	if k.isClosed() {
+		return
+	}
+
+	now := k.Clock.Now("CryptoKeyCache", "refresh")
+	k.fetchLock.Lock()
+	defer k.fetchLock.Unlock()
+
+	if k.isClosed() {
+		return
+	}
+
+	k.keysMu.RLock()
+	lastFetch := k.lastFetch
+	k.keysMu.RUnlock()
+
+	// There's a window we must account for where the timer fires while a fetch
+	// is ongoing but prior to the timer getting reset. In this case we want to
+	// avoid double fetching.
+	if now.Sub(lastFetch) < time.Minute*10 {
+		return
+	}
+
+	_, _, err := k.fetch(k.refreshCtx)
+	if err != nil {
+		k.logger.Error(k.refreshCtx, "fetch crypto keys", slog.Error(err))
+		return
+	}
+}
+
+// cryptoKeys queries the control plane for the crypto keys.
+// Outside of initialization, this should only be called by fetch.
+func (k *CryptoKeyCache) cryptoKeys(ctx context.Context) (map[int32]codersdk.CryptoKey, codersdk.CryptoKey, error) {
+	keys, err := k.client.CryptoKeys(ctx)
+	if err != nil {
+		return nil, codersdk.CryptoKey{}, xerrors.Errorf("crypto keys: %w", err)
+	}
+	cache, latest := toKeyMap(keys.CryptoKeys, k.Clock.Now())
+	return cache, latest, nil
+}
+
+// fetch fetches the keys from the control plane and updates the cache. The fetchMu
+// must be held when calling this function to avoid multiple concurrent fetches.
+// The returned keys are safe to use without additional locking.
+func (k *CryptoKeyCache) fetch(ctx context.Context) (map[int32]codersdk.CryptoKey, codersdk.CryptoKey, error) {
+	keys, latest, err := k.cryptoKeys(ctx)
+	if err != nil {
+		return nil, codersdk.CryptoKey{}, xerrors.Errorf("fetch keys: %w", err)
+	}
+
+	if len(keys) == 0 {
+		return nil, codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	now := k.Clock.Now()
+	if !latest.CanSign(now) {
+		return nil, codersdk.CryptoKey{}, cryptokeys.ErrKeyInvalid
```

Comment:
Rare, but if you're fetching to get a verification key, it seems wrong to throw an error based on the signing key.